### PR TITLE
Some node finder refactoring and optimization

### PR DESCRIPTION
**Pull request by danielhlarkin**
### Scope & Purpose

Some refactoring to reduce code duplication and reduce dynamic memory allocations in the optimizer.

Enterprise component: https://github.com/arangodb/enterprise/pull/514

- [x] Strictly *new functionality* (i.e. a new feature / new option, no need for porting)

### Testing & Verification

This change is a trivial rework / code cleanup without any test coverage.

Jenkins: https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/11127/
Jenkins Performance: http://172.16.10.101:8080/job/helper-perf-singleserver-simple-linux/1543/

### CODE REVIEW

File: arangod/Aql/CalculationNodeVarFinder.h [hunk 1]
```diff
@@ -30,7 +30,7 @@
 namespace arangodb {
 namespace aql {
 
-class CalculationNodeVarFinder final : public WalkerWorker<ExecutionNode> {
+class CalculationNodeVarFinder final : public WalkerWorker<ExecutionNode, false> {
```
Comment (jsteemann) [inline arangod/Aql/CalculationNodeVarFinder.h:33]: Although `true` / `false` works here, I find this harder to read/understand than necessary.
Can't we use something like `WalkerWorkerType::Unique` and `WalkerWorkerType::NonUnique` here? That would help me a lot looking at code.
That should also help against mindless copying&pasting code in the future.

File: arangod/Aql/ExecutionPlan.cpp [hunk 2]
```diff
@@ -2131,47 +2137,39 @@ ExecutionNode* ExecutionPlan::fromNode(AstNode const* node) {
...
-                                    ExecutionNode::NodeType type, bool enterSubqueries) {
-  // consult our nodes-of-type counters array
-  if (contains(type)) {
-    // node type present in plan
-    NodeFinder<ExecutionNode::NodeType> finder(type, result, enterSubqueries);
+                                    std::initializer_list<ExecutionNode::NodeType> const& types,
+                                    bool enterSubqueries) {
+  // check if any of the node types is actually present in the plan
+  Contains contains(*this);
+  bool haveNodes = std::any_of(types.begin(), types.end(), contains);
```
Comment (jsteemann) [inline arangod/Aql/ExecutionPlan.cpp:2147]: I have to say I find this less readable than a simple
```
bool haveNodes = std::any_of(types.begin(), types.end(), [this](ExecutionNode::NodeType type) {
  return contains(type);
});
```
But with the separate `Contains` I have to look elsewhere what it actually does.

File: arangod/Aql/CalculationNodeVarFinder.h (hunk 1 above)
Comment (danielhlarkin) [reply arangod/Aql/CalculationNodeVarFinder.h:33]: Ah, that's a good idea. I'll change that.

File: arangod/Aql/ExecutionPlan.cpp (hunk 2 above)
Comment (jsteemann) [reply arangod/Aql/ExecutionPlan.cpp:2147]: `this` should just cost 8 bytes, and with the g++ implementation lambdas should use an SSO with a 16 bytes buffer. So you could even capture two pointer values without getting into the dynamic memory allocation case.
I am not 100% sure about MSVC, but I guess it has a reasonable optimization for this common case too.

Comment (danielhlarkin) [reply arangod/Aql/ExecutionPlan.cpp:2147]: Okay, then I'll go ahead and convert to the lambda. I agree that it is definitely more readable.

Comment (jsteemann) [reply arangod/Aql/ExecutionPlan.cpp:2147]: Btw., what I wrote is true for `std::function`. It seems for lambdas the compiler is even smarter. Need to compile a bit more data...

Comment (mpoeter) [reply arangod/Aql/ExecutionPlan.cpp:2147]: `any_of` & co take the functor as template parameter by value, so the anonymous struct of the lambda is copied/moved into the function, so no dynamic allocation is necessary.

**Comment by danielhlarkin** [comment]
https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/11158/

**Comment by jsteemann** [comment]
https://172.16.10.101/view/PR/job/arangodb-matrix-pr/11159/

File: arangod/Aql/ExecutionPlan.cpp (hunk 2 above)
Comment (mpoeter) [reply arangod/Aql/ExecutionPlan.cpp]: FWIW, with respect to compiler optimization, the lambda is actually preferable. In `std::any_of(types.begin(), types.end(), contains)` the third parameter `contains` decays to a function pointer, but a function pointer is much harder to inline. With the lambda, the full type information of the (anonymous)  functor is available to the compiler, and that type information carries with it the code needed to be inlined.

**Pull request merged**